shared/src/types/Post.ts

```diff
@@ -1,28 +1,43 @@
-import type {Timestamp} from 'firebase-admin/firestore';
-import {LANGUAGE_TAG} from '../constants/i18n';
-import {UserProfile} from './User';
+import * as yup from 'yup';
+import {
+  DEFAULT_LANGUAGE_TAG,
+  LANGUAGE_TAG,
+  LANGUAGE_TAGS,
+} from '../constants/i18n';
+import {transformTimestamp} from '../modelUtils/transform';
+import {UserSchema} from './User';
 
-export type PostFields = {
-  id: string;
-  exerciseId: string;
-  sharingId: string;
-  userId: string | null;
-  language: LANGUAGE_TAG;
-  approved: boolean;
-  text: string;
-};
+const PostFiledsSchema = yup.object({
+  id: yup.string().required(),
+  exerciseId: yup.string().required(),
+  sharingId: yup.string().required(),
+  userId: yup.string().nullable().default(null),
+  language: yup.mixed<LANGUAGE_TAG>().oneOf(LANGUAGE_TAGS).required(),
+  approved: yup.boolean().required(),
+  text: yup.string().required(),
+});
+export type PostFields = yup.InferType<typeof PostFiledsSchema>;
 
-export type Post = PostFields & {
-  createdAt: string;
-  updatedAt: string;
-  userProfile?: UserProfile;
-};
+export const PostSchema = yup
+  .object({
+    createdAt: transformTimestamp.required(),
+    updatedAt: transformTimestamp.required(),
+    userProfile: yup.object().concat(UserSchema).nullable(),
+  })
+  .concat(PostFiledsSchema);
+export type Post = yup.InferType<typeof PostSchema>;
 
-export type PostData = PostFields & {
-  createdAt: Timestamp;
-  updatedAt: Timestamp;
-};
-
-export type PostParams = Omit<PostFields, 'id' | 'userId' | 'approved'> & {
-  anonymous: boolean;
-};
+export const CreatePostSchema = PostSchema.pick([
+  'exerciseId',
+  'sharingId',
+  'text',
+]).concat(
+  yup.object({
+    anonymous: yup.boolean().default(true),
+    language: yup
+      .mixed<LANGUAGE_TAG>()
+      .oneOf(LANGUAGE_TAGS)
+      .default(DEFAULT_LANGUAGE_TAG),
+  }),
+);
+export type CreatePostData = yup.InferType<typeof CreatePostSchema>;
```

Comment:
Nit pick, but maybe these should be called `CreatePostDataType` or `CreatePostType` just to not confuse them with schemas. Only if you agree.